agent/dockerclient/sdkclientfactory/sdkclientfactory.go

```diff
@@ -0,0 +1,193 @@
+	} else {
+		return nil, errors.New("docker client factory: client not found for docker version: " + string(version))
+	}
+}
+
+// findDockerVersions loops over all known API versions and finds which ones
+// are supported by the docker daemon on the host
+func findDockerVersions(ctx context.Context, endpoint string) map[dockerclient.DockerVersion]sdkclient.Client {
+	// if the client version returns a MinAPIVersion, use it to return all the
+	// Docker clients between MinAPIVersion and APIVersion, else try pinging
+	// the clients in getKnownAPIVersions
+	var minAPIVersion, apiVersion string
+	// get a Docker client with the default supported version
+	client, err := newVersionedClient(endpoint, string(minDockerAPIVersion))
+	if err == nil {
+		derivedCtx, cancel := context.WithTimeout(ctx, dockerclient.VersionTimeout)
+		defer cancel()
+		serverVersion, err := client.ServerVersion(derivedCtx)
+		if err == nil {
+			apiVersion = serverVersion.APIVersion
+			// check if the docker.Version has MinAPIVersion value
+			if serverVersion.MinAPIVersion != "" {
+				minAPIVersion = serverVersion.MinAPIVersion
+			}
+		}
+	}
+	clients := make(map[dockerclient.DockerVersion]sdkclient.Client)
+	for _, version := range dockerclient.GetKnownAPIVersions() {
+		dockerClient, err := getDockerClientForVersion(endpoint, string(version), minAPIVersion, apiVersion)
+		if err != nil {
+			log.Infof("Unable to get Docker client for version %s: %v", version, err)
+			continue
+		}
+		clients[version] = dockerClient
+	}
+	return clients
+}
+
+func getDockerClientForVersion(
+	endpoint string,
+	version string,
+	minAPIVersion string,
+	apiVersion string) (sdkclient.Client, error) {
+	if minAPIVersion != "" && apiVersion != "" {
+		// Adding patch number zero to Docker versions to reuse the existing semver
+		// comparator
+		// TODO: remove this logic later when non-semver comparator is implemented
+		versionWithPatch := version + zeroPatch
+		lessThanMinCheck := "<" + minAPIVersion + zeroPatch
+		moreThanMaxCheck := ">" + apiVersion + zeroPatch
+		minVersionCheck, err := utils.Version(versionWithPatch).Matches(lessThanMinCheck)
+		if err != nil {
+			return nil, errors.Wrapf(err, "version detection using MinAPIVersion: unable to get min version: %s", minAPIVersion)
+		}
+		maxVersionCheck, err := utils.Version(versionWithPatch).Matches(moreThanMaxCheck)
+		if err != nil {
+			return nil, errors.Wrapf(err, "version detection using MinAPIVersion: unable to get max version: %s", apiVersion)
+		}
+		// Do not add the version when it is outside minVersion to maxVersion
+		if minVersionCheck || maxVersionCheck {
+			return nil, errors.Errorf("version detection using MinAPIVersion: unsupported version: %s", version)
+		}
+	}
+	client, err := newVersionedClient(endpoint, string(version))
+	if err != nil {
+		return nil, errors.Wrapf(err, "version detection check: unable to create Docker client for version: %s", version)
+	}
+	_, err = client.Ping(context.TODO())
```

Comment:
should we have `getDockerClientForVersion` take a context? so you can pass in a derived context from `findDockerVersions` instead of the `context.TODO`?